### PR TITLE
REECB Documentation + Implementation

### PR DESCRIPTION
**Pull request by lukelowry**
## Description

Implementation and documentation for the `REECB` PhasorDynamics converter model.

## Proposed changes

- Added `REECB` model implementation, documentation, and diagram.
- Added `REECB` JSON parsing and `SystemModel` assembly support.
- Added focused `PhasorDynamicsConverterReecbTest` unit coverage.
- Added minimal CommonMath compatibility needed for REECB dependency tracking/Jacobian paths.
- Updated converter registration, input-format docs, and changelog.

## Checklist

- [x] All tests pass (except known PE failure)
- [x] Code compiles cleanly with flags `-Wall -Wpedantic -Wconversion -Wextra`.
- [x] The new code follows GridKit™ style guidelines.
- [x] There are unit tests for the new code.
- [x] The new code is documented.
- [x] The feature branch is rebased with respect to the target branch.
- [x] I have updated [CHANGELOG.md](/CHANGELOG.md) to reflect the changes in this PR.

## Further comments

Intended to be merged after #436

### CODE REVIEW

**Comment by pelesh** [comment]
Needs rebasing

File: tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp [hunk 1]
```diff
@@ -716,6 +721,170 @@ namespace GridKit
...
+      static constexpr IdxT kPrefSignalId          = static_cast<IdxT>(205);
+      static constexpr IdxT kConverterComponentId  = static_cast<IdxT>(0);
+      static constexpr IdxT kControllerComponentId = static_cast<IdxT>(1);
+
+      static constexpr RealT kReferencePulse      = static_cast<RealT>(0.05);
+      static constexpr RealT kPulseEnd            = static_cast<RealT>(0.1);
+      static constexpr RealT kResponseTolerance   = static_cast<RealT>(0.01);
+      static constexpr RealT kRecoveryHorizon     = static_cast<RealT>(25.0);
+      static constexpr RealT kRecoveryMonitorStep = static_cast<RealT>(1.0 / 60.0);
+      static constexpr RealT kRecoveryTolerance   = static_cast<RealT>(1.0e-6);
```
Comment (nkoukpaizan) [inline tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp]: If these are only used by `regcaReecbRecovery`, can we define them there instead of as private member variables to `PDIntegrationTests`?

File: tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp [hunk 2]
```diff
@@ -716,6 +721,170 @@ namespace GridKit
...
+        using RegcaVar = RegcaInternalVariables;
+
+        TestStatus success = true;
+
+        SystemModelDataT data;
+        data.va_base = static_cast<RealT>(100.0e6);
+
+        auto& bus    = data.bus.emplace_back();
+        bus.bus_id   = kReecbBusId;
+        bus.bus_type = BusDataT::BusType::SLACK;
```
Comment (nkoukpaizan) [inline tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp:779]: `[WARNING] A component has returned a nullptr Jacobian.`
Is this the desired behavior, or can we use `BusType::DEFAULT` here.

File: tests/UnitTests/PhasorDynamics/SystemSingleComponentTests.hpp [hunk 3]
```diff
@@ -252,6 +253,101 @@ namespace GridKit
...
+        success *= system.size()
+                   == static_cast<IdxT>(RegcaVars::MAXIMUM)
+                          + static_cast<IdxT>(Vars::MAXIMUM);
+
+        return success.report(__func__);
+      }
+
+      /// System initialization reports a statically valid component whose
+      /// operating point cannot be initialized.
+      TestOutcome initializationFailure()
```
Comment (nkoukpaizan) [inline tests/UnitTests/PhasorDynamics/SystemSingleComponentTests.hpp]: Is this specific to `REGCA`? If so, I recommend a more meaningful name for the test. If not, consider moving to `SystemTests.hpp`.

File: GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp [hunk 4]
```diff
@@ -0,0 +1,1320 @@
...
+        const ScalarT ipcmd  = toComponentBase(ipcmd_system);
+
+        const ScalarT vmeas_safe = Math::max(vmeas, VMEAS_MINIMUM);
+        const ScalarT sdip       = Math::inside(vt, Vdip_, Vup_);
+        const ScalarT verr       = Math::deadband2(Vref0_ - vmeas, dbd1_, dbd2_);
+        const ScalarT iqv        = Math::clamp(kqv_ * verr, Iql1_, Iqh1_);
+        // The Volt/VAr channel is a system-base reactive power unless
+        // direct-voltage mode selects it as a terminal-voltage reference,
+        // which takes no power-base conversion.
+        const ScalarT qref       = q_ref_on_ * (pf_on_ * pmeas * std::tan(pfaref) + pf_off_ * toComponentBase(extref));
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: There is an issue how Enzyme handles `std::tan` with `Clang 19` and newer. I've created an [issue on the Enzyme repo](https://github.com/EnzymeAD/Enzyme/issues/3101). For now you can add the `-fno-builtin-tan` tag to the compile flags for this model or rewrite as `sin/cos`.

**Review by nkoukpaizan — APPROVED** [review]
Overall looks good. 

A few minor comments. I can't load the diff online, but I think the Jacobian tests will need to be updated to avoid handling missing entries and zeros differently (as discussed in #422.

**Comment by lukelowry** [comment]
Will rebase and address these issues. will be done soon. this is super super helpful for validation thank you. we are on a roll

File: GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp [hunk 5]
```diff
@@ -1103,9 +1191,9 @@ namespace GridKit
        */
       template <typename scalar_type, typename index_type>
       typename Reecb<scalar_type, index_type>::RealT
-      Reecb<scalar_type, index_type>::logOneMinusExp(RealT x) const
+      Reecb<scalar_type, index_type>::logOneMinusExp(RealT x)
       {
-        static constexpr RealT log_two = std::numbers::ln2_v<RealT>;
+        static const RealT log_two = std::log(TWO<RealT>);
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: This seems like a recent regression.

**Review by nkoukpaizan — COMMENTED** [review]
Something happened in the past few commits. I'm now getting extremely long build times on one of my testing backends. 

Flagging one regression I've seen, though that would not be the cause of compilation issues.

File: GridKit/Model/PhasorDynamics/SystemModelImpl.hpp [hunk 6]
```diff
@@ -690,21 +749,28 @@ namespace GridKit
...
+        }
         evaluateResidual();
         evaluateJacobian();
       }
 
+      // Start variable monitors only after allocation and sparse initialization
+      // complete successfully.
+      initializeMonitor();
+      startMonitor();
+
```
Comment (pelesh) [inline GridKit/Model/PhasorDynamics/SystemModelImpl.hpp:883]: Why this change? Is there a related issue? If not, either open an issue or document better.

File: GridKit/Model/PhasorDynamics/Controller/REECB/Reecb.hpp [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+      private:
+        /// Smooth asymmetric slew-rate limiter.
+        [[gnu::always_inline]] static inline ScalarT aslew(ScalarT rate, RealT lower, RealT upper);
+
+        /// Smooth anti-windup derivative within a moving symmetric band.
+        [[gnu::always_inline]] static inline ScalarT awband(ScalarT state, ScalarT rate, ScalarT band);
+
+        static void checkConfiguration(bool condition, const char* message, int& errors);
+        void        loadRealParameter(const ModelDataT& data,
```
Comment (pelesh) [inline GridKit/Model/PhasorDynamics/Controller/REECB/Reecb.hpp]: Defining static method `checkConfiguration` looks more like self-documenting code than using lambdas. Might be good idea to update other component models accordingly (not in this PR, though) or if we want to stick with lambdas, align this component model with others. Consistency is important.

File: GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp [hunk 8]
```diff
@@ -0,0 +1,1320 @@
...
+
+        RealT ilrhs0  = ZERO<RealT>;
+        RealT ilmax0  = ZERO<RealT>;
+        RealT ilnorm0 = ZERO<RealT>;
+        RealT ilcap0  = ZERO<RealT>;
+        for (int correction = 0; correction <= std::numeric_limits<RealT>::digits && std::isfinite(imax); ++correction)
+        {
+          ilrhs0 = (imax - high0) * (imax + high0);
+          ilmax0 = ZERO<RealT>;
+          if (ilrhs0 > ZERO<RealT>)
```
Comment (pelesh) [inline GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: Claude asks: Can we use bisection search as in HYGOV?

Comment (pelesh) [reply GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: In any case, this loop needs to be commented better because it is unclear what is happening here.

**Review by pelesh — APPROVED** [review]
Looks good to me. 

Some nitpicking comments below.

File: tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp (hunk 1 above)
Comment (lukelowry) [reply tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp]: I removed and cleaned. This was bad anyway. Thank you

File: GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp (hunk 5 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: I didn't realize this would affect compile time. I misunderstood the purpose of `ln2_v`, sorry about that.

Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: Oh and to clarify that semeed to speed up my compilation so I am not certain but I dont want to kill more time on that since its fast again for me

File: GridKit/Model/PhasorDynamics/Controller/REECB/Reecb.hpp (hunk 7 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Controller/REECB/Reecb.hpp]: I changed to `check` to be more consistent. We can extract later, so we dont have to rewrite this each time

File: GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp (hunk 8 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: I extracted a `bisect` function to make the initialization cleaner, and it's used in a different way from `HYGOV,` but it's much more readable/documented now. thank you

File: GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp (hunk 4 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: Added `-fno-builtin-tan` for `phasor_dynamics_controller_reecb` as requested and I kept the model equation using `std::tan`. amazing find

File: GridKit/Model/PhasorDynamics/SystemModelImpl.hpp (hunk 6 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/SystemModelImpl.hpp:883]: This prevents sparse Jacobian discovery from evaluating an invalid operating point when a component fails to initialize. I documented that requirement in `allocate()` and added `componentInitializationError()` to verify that the failure is propagated before residual/Jacobian evaluation or monitor startup.

I started running into this issue now that I am in a better place to do full validation tests, especially since I have been a lot stricter with parameter validation in the newer models.

File: tests/UnitTests/PhasorDynamics/SystemSingleComponentTests.hpp (hunk 3 above)
Comment (lukelowry) [reply tests/UnitTests/PhasorDynamics/SystemSingleComponentTests.hpp]: I moved it to `SystemTests.hpp` as `componentInitializationError()` and replaced the REGCA setup with a generic component that fails initialization. The remaining `reecb()`  is REECB-only.

File: tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp (hunk 2 above)
Comment (lukelowry) [reply tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp:779]: The warning came from `ConstantSignalSource`, not the `SLACK` bus. It has an empty Jacobian, but previously left the Jacobian pointer null. Its `evaluateJacobian()` now constructs a valid zero-entry COO matrix, so the warning is gone. The `SLACK` bus is intentional to provide the fixed terminal voltage for this test.

**Comment by lukelowry** [comment]
All, this is ready. This is by far the most complicated model we have implemented. I didn't originally think this would be possible with a fully smooth Jacobian, so this is an important milestone demonstrating the feasibility for even larger inverter/rectifier models. Thank you to @nkoukpaizan for your patience and diligent reviews

**Review by abirchfield — APPROVED** [review]

File: GridKit/Model/PhasorDynamics/Controller/REECB/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    COMPILE_OPTIONS
+      PRIVATE
+      -mllvm
+      -enzyme-auto-sparsity=1
+      -fno-math-errno)
+
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 19)
+    # Work around EnzymeAD/Enzyme#3101.
+    target_compile_options(phasor_dynamics_controller_reecb PRIVATE -fno-builtin-tan)
+  endif()
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Controller/REECB/CMakeLists.txt]:
```suggestion

```
Suggesting adding the flag to the macro call directly. I would also avoid including GitHub issue numbers in the repo itself. I've found this hard to maintain in the past.

File: GridKit/Model/PhasorDynamics/Controller/REECB/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      GridKit::phasor_dynamics_core
+      PUBLIC
+      GridKit::phasor_dynamics_signal
+      PRIVATE
+      ClangEnzymeFlags
+    COMPILE_OPTIONS
+      PRIVATE
+      -mllvm
+      -enzyme-auto-sparsity=1
+      -fno-math-errno)
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Controller/REECB/CMakeLists.txt]:
```suggestion
      -fno-math-errno
      -fno-builtin-tan) 
```

File: GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp [hunk 11]
```diff
@@ -0,0 +1,1468 @@
...
+
+        const RealT result      = bisect(a, b, below);
+        const RealT final_state = circleState(result, high);
+        const RealT final_cap   = capacity(final_state);
+        if (!std::isfinite(result) || !std::isfinite(final_state)
+            || !std::isfinite(final_cap) || final_cap < low)
+        {
+          return nan;
+        }
+        return result;
```
Comment (nkoukpaizan) [inline GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: I think this needs to be further documented. I don't really understand what's going on based on the method's documentation.

File: tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp [hunk 12]
```diff
@@ -716,6 +723,222 @@ namespace GridKit
...
+        constexpr IdxT  RENEWABLE_BUS_ID       = static_cast<IdxT>(23);
+        constexpr IdxT  IPCMD_SIGNAL_ID        = static_cast<IdxT>(201);
+        constexpr IdxT  IQCMD_SIGNAL_ID        = static_cast<IdxT>(202);
+        constexpr IdxT  IBRANCHR_SIGNAL_ID     = static_cast<IdxT>(203);
+        constexpr IdxT  IBRANCHI_SIGNAL_ID     = static_cast<IdxT>(204);
+        constexpr IdxT  PBRANCH_SIGNAL_ID      = static_cast<IdxT>(205);
+        constexpr IdxT  QBRANCH_SIGNAL_ID      = static_cast<IdxT>(206);
+        constexpr IdxT  QEXT_SIGNAL_ID         = static_cast<IdxT>(207);
+        constexpr IdxT  PEXT_SIGNAL_ID         = static_cast<IdxT>(208);
+        constexpr IdxT  PLANT_PREF_SIGNAL_ID   = static_cast<IdxT>(209);
```
Comment (nkoukpaizan) [inline tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp]: Thinking about maintaining this later, could you clarify the rationale for these constants? Could I set any unique numbers? Perhaps add some comments.

File: tests/UnitTests/PhasorDynamics/ControllerReecbTests.hpp [hunk 13]
```diff
@@ -0,0 +1,2681 @@
...
+                success *= derivativeMatches(dependency, Vars::VT, kBusVrColumn, 1.8, "VT-Vr");
+                success *= derivativeMatches(dependency, Vars::VT, kBusViColumn, 0.8, "VT-Vi");
+                success *= derivativeMatches(dependency, Vars::ILMAX, Vars::ILMAX, -4.0, "ILMAX diagonal");
+                success *= derivativeMatches(dependency, Vars::IQCMD, Vars::IQCMD, -2.0, "IQCMD diagonal");
+                success *= derivativeMatches(dependency, Vars::IPCMD, Vars::IPCMD, -2.0, "IPCMD diagonal");
+                success *= derivativeMatches(dependency, Vars::IPCMD, Vars::PORD, 1.0, "IPCMD-PORD");
+                success *= derivativeMatches(dependency, Vars::IPCMD, Vars::VMEAS, -0.5, "IPCMD-VMEAS");
+                success *= derivativeMatches(dependency, Vars::IQCMD, Vars::XPIV, reactive ? 1.0 : 0.0, "IQCMD-XPIV selector path");
+                success *= derivativeMatches(dependency, Vars::IQCMD, Vars::QV, reactive ? 0.0 : 1.0, "IQCMD-QV selector path");
+                success *= derivativeMatches(
```
Comment (nkoukpaizan) [inline tests/UnitTests/PhasorDynamics/ControllerReecbTests.hpp]: I would create the reference map and compare the maps directly. 

I am also questioning the value of this kind of tests for large models. If the model changes, we'll need to update the reference values for these tests. Comparing two automatic differentiation methods (dependency tracking and Enzyme) lets us avoid that. There are other tests to verify those methods against analytical solutions on smaller models. At some point, it's okay to not test every possible edge case if the maintainability burden is too high.

**Review by nkoukpaizan — COMMENTED** [review]
A few more comments before we merge this.

File: GridKit/Model/PhasorDynamics/Controller/REECB/CMakeLists.txt (hunk 9 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Controller/REECB/CMakeLists.txt]: Suggestion applied, thank you

File: tests/UnitTests/PhasorDynamics/ControllerReecbTests.hpp (hunk 13 above)
Comment (pelesh) [reply tests/UnitTests/PhasorDynamics/ControllerReecbTests.hpp]: We have `DependencyTracking` automatic differentiation, which is probably more reliable test for automatic differentiation than the hand-computed Jacobian. `DependencyTracking` is well tested and requires only C++03 standard.

Comment (lukelowry) [reply tests/UnitTests/PhasorDynamics/ControllerReecbTests.hpp]: Okay, I have addressed this. I thought that eliminating the algebraic variables would help with runtime speed, but I believe that is what contributed partially to the long build time. I believe this is resolved now.

We can remove the excessive tests once we become a bit more confident everything is working ideally

File: tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp (hunk 12 above)
Comment (lukelowry) [reply tests/IntegrationTests/PhasorDynamics/PDIntegrationTests.hpp]: Changed to be more consistent. We definitely need better fixtures or not have signals as independent 'nodes'. Regardless, I cleaned this up. much simpler now I think

File: GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp (hunk 11 above)
Comment (lukelowry) [reply GridKit/Model/PhasorDynamics/Controller/REECB/ReecbImpl.hpp]: Updated the documentation and split the calculation into named steps. (1) Find a feasible upper current limit; (2) bisect to the smallest feasible value; and  (3)return its current-circle state and remaining capacity.

I really do apologize for the complexity. this is more or less how it's documented in the model specifications I am referencing. I added footnote in [7c4316b](https://github.com/ORNL/GridKit/pull/421/commits/7c4316bd59e1b57900a42f6bcd0077367443d263) to help future readers

**Comment by lukelowry** [comment]
Alright, @nkoukpaizan, let me know what you think. We can do some back and forth if you have any further concerns. Ill watch this thread closely

**Comment by pelesh** [comment]
@lukelowry please rebase and we can merge. Great work!

**Comment by lukelowry** [comment]
@pelesh rebased just waiting on CI
